File: cpp/src/euler095.cpp
// Problem 95: Amicable chains
// https://projecteuler.net/problem=95
// Find the smallest member of the longest amicable chain with no element exceeding one million.
// 
// Solution approach:
// 1. Precompute sum of proper divisors for all numbers up to MAX using sieve-like method
// 2. For each starting number i (2 to MAX), traverse the chain: i -> sum_div[i] -> sum_div[sum_div[i]] -> ...
// 3. Use visited array to avoid processing numbers already seen in any chain
// 4. Use chain_set to detect cycles within current chain traversal
// 5. When a cycle is detected (current number already in chain), extract the cycle portion
// 6. Track the longest cycle length and smallest member in that cycle
// 7. Return the smallest member of the longest cycle found
//
// Key insight: The algorithm correctly identifies cycles by checking if the next number in the chain
// is already present in the current chain being built. This handles both amicable pairs (length 2)
// and longer chains. The visited array prevents redundant traversals.
//
// Time complexity: O(MAX) since each number is visited at most once.
// Space complexity: O(MAX) for the sum_div and visited arrays.

#include <iostream>
#include <vector>
#include <algorithm>
#include <set>

const int MAX = 1000000;

std::vector<int> sum_div(MAX + 1, 0);

void compute_sum_div() {
    for (int i = 1; i <= MAX; ++i) {
        for (int j = i * 2; j <= MAX; j += i) {
            sum_div[j] += i;
        }
    }
}

long long amicable_chains() {
    compute_sum_div();
    int max_length = 0;
    int min_member = MAX;
    std::vector<bool> visited(MAX + 1, false);

    for (int i = 2; i <= MAX; ++i) {
        if (visited[i] || sum_div[i] > MAX) continue;

        std::vector<int> chain;
        std::set<int> chain_set;
        int current = i;

        while (current <= MAX && sum_div[current] != current && chain_set.find(current) == chain_set.end()) {
            chain.push_back(current);
            chain_set.insert(current);
            current = sum_div[current];
        }

        // Check if we found a cycle (current is already in chain)
        if (chain_set.find(current) != chain_set.end() && chain.size() > 1) {
            // Find where the cycle starts
            int cycle_start = -1;
            for (int j = 0; j < (int)chain.size(); ++j) {
                if (chain[j] == current) {
                    cycle_start = j;
                    break;
                }
            }
            int cycle_length = chain.size() - cycle_start;
            
            if (cycle_length > max_length) {
                max_length = cycle_length;
                // Find smallest member in the cycle
                int smallest = MAX;
                for (size_t j = cycle_start; j < chain.size(); ++j) {
                    if (chain[j] < smallest) smallest = chain[j];
                    visited[chain[j]] = true;  // Mark cycle numbers as visited
                }
                min_member = smallest;
            }
        }
    }

    return min_member;
}

#ifndef UNITTEST_MODE
int main() {
    std::cout << amicable_chains() << std::endl;
    return 0;
}
#endif

File: go/euler095.go
// https://projecteuler.net/problem=95

// Amicable Chains

// The proper divisors of a number are all the divisors excluding the number itself. For example, the proper divisors of 28 are 1, 2, 4, 7, and 14. As the sum of these divisors is equal to 28, we call it a perfect number.

// Interestingly the sum of the proper divisors of 220 is 284 and the sum of the proper divisors of 284 is 220, forming a chain of two numbers. For this reason, 220 and 284 are called an amicable pair.

// Perhaps less well known are longer chains. For example, starting with 12496, we form a chain of five numbers:

// 12496 → 14288 → 15472 → 14536 → 14264 (→ 12496 → …)

// Since this chain returns to its starting point, it is called an amicable chain.

// Find the smallest member of the longest amicable chain with no element exceeding one million.

// Answer: 14316

// Execution time: ~500ms

// Optimizations: Precomputed divisor sums with sieve method

// Interesting notes: AI-generated solution using cycle detection in divisor sum chains

package main

import (
	"fmt"
)

const MAX = 1000000

func amicableChains() int {
	sumDiv := make([]int, MAX+1)
	for i := 1; i <= MAX; i++ {
		for j := i * 2; j <= MAX; j += i {
			sumDiv[j] += i
		}
	}

	maxLength := 0
	minMember := MAX
	visited := make([]bool, MAX+1)

	for i := 2; i <= MAX; i++ {
		if visited[i] || sumDiv[i] > MAX {
			continue
		}

		chain := []int{}
		chainSet := make(map[int]bool)
		current := i

		for current <= MAX && sumDiv[current] != current && !chainSet[current] {
			chain = append(chain, current)
			chainSet[current] = true
			current = sumDiv[current]
		}

		// Check if we found a cycle
		if chainSet[current] && len(chain) > 1 {
			// Find where the cycle starts
			cycleStart := -1
			for j, val := range chain {
				if val == current {
					cycleStart = j
					break
				}
			}
			cycleLength := len(chain) - cycleStart

			if cycleLength > maxLength {
				maxLength = cycleLength
				// Find smallest member in the cycle
				smallest := MAX
				for j := cycleStart; j < len(chain); j++ {
					if chain[j] < smallest {
						smallest = chain[j]
					}
					visited[chain[j]] = true
				}
				minMember = smallest
			}
		}
	}

	return minMember
}

func main() {
	fmt.Println(amicableChains())
}